src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java

```diff
@@ -81,6 +81,23 @@ public ElementNode(AbstractNode root, AbstractNode parent, DetailAST detailAst)
         createChildren();
     }
 
+    /**
+     * Compares current object with specified for order.
+     * @param other another {@code NodeInfo} object
+     * @return number representing order of current object to specified one
+     */
+    @Override
+    public int compareOrder(NodeInfo other) {
```

Comment:
@gaurabdg @timurt How often is `compareOrder` called for different queries? Is it a big impact that we have to calculate the depth each time it is called?

Reply:
@rnveach I can kind of cache the depth value for each node, as the two compared nodes least common ancestor is very near(because this ordering is required for xpath queries for nearby nodes mostly)  and will save a lot of computations.

Reply:
Done

Reply:
So I was right with my concern that it is called many times for 1 query?

Reply:
Yes, it will be called for each xpath result ordering

Reply:
The order of cousins will be wrong. Consider
```
CLASS_DEF
   METHOD_DEF1
       VARIABLE_DEF1-1
       VARIABLE_DEF1-2
   METHOD_DEF2
       VARIABLE_DEF2-1
       VARIABLE_DEF2-2
```
The list of nodes for the xpath "//VARIABLE_DEF" will be VARIABLE_DEF1-1, VARIABLE_DEF2-1,VARIABLE_DEF2-2,VARIABLE_DEF1-2.

We still need some sort of recursion here. When the `depth` is different, the result is clear.
When the `depth` matches, there should be a check for order for the parent nodes (in case both nodes do not belong to the same parent).